### PR TITLE
Move FSDP module transformation to transform_module

### PR DESCRIPTION
**Pull request by t-vi**
This moves the modification of them Module inside `FSDPTransform.transform_module`.
There currently are no changes how it operates.

This is the first step towards a composable FSDP Transform, I filed #987 to discuss the broader plan.

cc @crcrpar @lantiga

### CODE REVIEW

File: thunder/distributed/tensor_parallel/column_wise.py [hunk 1]
```diff
@@ -284,6 +284,7 @@ def forward(self, tokens: torch.Tensor) -> torch.Tensor:
             chunked_param_name_to_layer_type=chunked_param_name_to_layer_type,
             process_group=process_group,
         ),
+        _legacy_copy_params=True,
```
Comment (crcrpar) [inline thunder/distributed/tensor_parallel/column_wise.py:287]: I'd have tensor parallel transforms behave in the same way as fsdp v2 in regard to param handling; direct sharding w/o copies.

File: thunder/distributed/tensor_parallel/row_wise.py [hunk 2]
```diff
@@ -291,6 +291,7 @@ def forward(self, tokens: torch.Tensor) -> torch.Tensor:
             chunked_param_name_to_layer_type=chunked_param_name_to_layer_type,
             process_group=process_group,
         ),
+        _legacy_copy_params=True,
```
Comment (crcrpar) [inline thunder/distributed/tensor_parallel/row_wise.py:294]: same comment as column_wise.py

File: thunder/distributed/tensor_parallel/column_wise.py (hunk 1 above)
Comment (crcrpar) [reply thunder/distributed/tensor_parallel/column_wise.py:287]: also, if we go with this parameter being `True`, by default, which transforms want to copy params?

File: thunder/distributed/transforms/fsdp_v2.py [hunk 3]
```diff
@@ -57,12 +65,166 @@ def __call__(self, bsym: BoundSymbol) -> VISIT_TYPE:
         return VISIT_TYPE.REPLACE
 
 
-@dataclass(frozen=True)
-class FSDPTraceTransform(Transform):
+# When the user calls fsdp(jitted_module), this function does the following
+# - It transforms the ThunderModule jitted_module, materializing and sharding the parameters as `overrides`
+#   in the ThunderModule.
+# - While doing that, it leaves the original user module alone.
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:71]: is this comment still true?

File: thunder/distributed/transforms/fsdp_v2.py [hunk 4]
```diff
@@ -57,12 +65,166 @@ def __call__(self, bsym: BoundSymbol) -> VISIT_TYPE:
...
+# - It then registers an early transform (callback that runs before prologue is executed) that transforms the
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:72]: is it still an early transform or a vanilla transform?

File: thunder/distributed/transforms/fsdp_v2.py [hunk 5]
```diff
@@ -57,12 +65,166 @@ def __call__(self, bsym: BoundSymbol) -> VISIT_TYPE:
...
+        self,
+        device: torch.device | None = None,
+        broadcast_from: int | None = None,
+        sharding_strategy: FSDPType = FSDPType.ZERO2,
+        bucketing_strategy: FSDPBucketingStrategy = FSDPBucketingStrategy.NONE,
+    ):
+        self.device = device
+        self.broadcast_from = broadcast_from
+        self.sharding_strategy = sharding_strategy
+        self.bucketing_strategy = bucketing_strategy
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:95]: nit-picking: I might prefer explicit init of these three here, and check the size / type of them in their callsite.
```suggestion
        self.bucketing_strategy = bucketing_strategy
        self.sharded_params: dict[str, Any] = {}
        self.process_group: ProcessGroup | None = None
        self.shared_params_name: dict[str, str] = {}
```

File: thunder/distributed/tensor_parallel/column_wise.py (hunk 1 above)
Comment (t-vi) [reply thunder/distributed/tensor_parallel/column_wise.py:287]: Yes, this parameter is just a "let's keep Tensor Parallel without changing the transform in this PR", but we would want to update tensor parallel in the same way.

File: thunder/distributed/transforms/fsdp_v2.py (hunk 4 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:72]:
```suggestion
# - It then registers an transform (callback that runs before prologue is executed) that transforms the
```

File: thunder/distributed/transforms/fsdp_v2.py (hunk 3 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:71]: Briefly, #988 will change the original user module if requested to free the unsharded parameters.

File: thunder/distributed/tensor_parallel/column_wise.py (hunk 1 above)
Comment (t-vi) [reply thunder/distributed/tensor_parallel/column_wise.py:287]: I hope to get to the fourth part (delegate materialization to MaterializationTransform) ideally tomorrow, then we can see if we want to split the tensor parallel in the same parts or different ones.

File: thunder/distributed/transforms/fsdp_v2.py (hunk 3 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:71]: I updated this in #995

File: thunder/distributed/__init__.py [hunk 6]
```diff
@@ -621,13 +472,19 @@ def fsdp(
     )
 
     if isinstance(model, thunder.ThunderModule):
-        return fsdp_transform_module(
+        from thunder.core.transforms import add_transform
+        from thunder.distributed.transforms.fsdp_v2 import FSDPTransform
+
+        new_model = add_transform(
```
Comment (crcrpar) [inline thunder/distributed/__init__.py:478]: nit-picking

```suggestion
        return add_transform(
```

File: thunder/distributed/__init__.py [hunk 7]
```diff
@@ -621,13 +472,19 @@ def fsdp(
...
-            sharding_strategy=sharding_strategy,
-            bucketing_strategy=bucketing_strategy,
+            transform=FSDPTransform(
+                device=device,
+                broadcast_from=broadcast_from,
+                sharding_strategy=sharding_strategy,
+                bucketing_strategy=bucketing_strategy,
+            ),
         )
+        return new_model
```
Comment (crcrpar) [inline thunder/distributed/__init__.py:487]:
```suggestion
```

**Review by crcrpar — APPROVED** [review]

**Comment by t-vi** [comment]
Thank you @crcrpar

**Pull request merged**